Add pull_policy to the spec (#59)
compose-spec/compose-spec: spec.md

    image: my_private.registry:5000/redis
```    

If the image does not exist on the platform, Compose implementations MUST attempt to pull it based on the `pull_policy`. Compose
implementations with build support MAY offer alternative options for the end user to control precedence of
pull over building the image from source, however pulling the image MUST be the default behavior.

`image` MAY be omitted from a Compose file as long as a `build` section is declared. Compose implementations
without build support MUST fail when `image` is missing from the Compose file.

### pull_policy

> `pull_policy` define the decisions Compose implementations will make when it starts to pull images. Possible values are:

Reviewer: ```suggestion
`pull_policy` defines the decisions Compose implementations will make when it starts to pull images. Possible values are:
```

> * `always`: Compose implementations SHOULD always pull the image whether it's been pulled or not.

Reviewer: ```suggestion
* `always`: Compose implementations SHOULD pull the image from the registry ignoring other caches.
```

> * `never`: Compose implementations SHOULD NOT pull the image, instead it SHOULD use the image on the local.

Reviewer: ```suggestion
* `never`: Compose implementations SHOULD NOT pull the image from a registry and SHOULD rely on the platform cached image. If there is no cached image, a failure MUST be reported. 
```

> * `if_not_present`: Compose implementations SHOULD pull the image only if it's not found on the local.This SHOULD be the default option for Compose implementations without build support.

Reviewer: ```suggestion
* `if_not_present`: Compose implementations SHOULD pull the image from the registry only if it's not available in the platform cache. This SHOULD be the default option for Compose implementations without build support.
```

> `pull_policy` MAY be omitted from a Compose file as long as a `build` section is declared.

Reviewer: We say higher up that the default is `if_not_present` so we need to make this language clear for that case.

What about if both `pull_policy` and `build` are declared?

Author: Thanks for all these great advices(Sorry about my written english...). I think allow user to specify both `build` and `pull_policy` is  not necessary, simply report an error should be a good choice.

Reviewer: No problem!

> I think allow user to specify both build and pull_policy is not necessary, simply report an error should be a good choice.

I can see a case where you would want both build and pull policy set: if you're building your images locally to testing them and then want deploy the same Compose file to a remote endpoint. Ideally you wouldn't have to edit the Compose file between those steps to replace the build section with a pull policy.

An imagined UX for this could be:
```console
$ docker-compose up --build # locally build images and run
$ docker-compose push # push all built images to a registry
$ docker-compose up --context remote # run remotely using pull policy
```

The above assumes that by default we try to pull images (respecting the pull policy) and have an override flag to build. This logic could be reversed and we may want different logic for different implementations.

A straw man that allows implementations to decide could be:
> If `pull_policy` and `build` are omitted then implementations MUST pull the image using the default pull policy of `if_not_present`. If both `build` and `pull_policy` are set then implementations MAY build the image or MAY pull it respecting the pull policy. If only `build` is set then implementations MAY build the image or pull it using the default pull policy.

Author: Seems a good idea to use cmd flags. But in the spec, i'm afraid the aboving description can be confused for users and implementations. Since we can use cmd flags to override the default behavior, can we difine the default behavior as `build > pull_policy` in the spec, and make it clear that implements can provide cli flags or other ways to override this.

I'm not sure if i made my point clear.

Reviewer: I like the idea of having a SHOULD in the spec and explicitly saying that implementations can override this behaviour

Author: Updated

    image: my_private.registry:5000/redis
```    

If the image does not exist on the platform, Compose implementations MUST attempt to pull it based on the `pull_policy`. Compose
implementations with build support MAY offer alternative options for the end user to control precedence of
pull over building the image from source, however pulling the image MUST be the default behavior.

`image` MAY be omitted from a Compose file as long as a `build` section is declared. Compose implementations
without build support MUST fail when `image` is missing from the Compose file.

### pull_policy

`pull_policy` defines the decisions Compose implementations will make when it starts to pull images. Possible values are:

> * `always`: Compose implementations SHOULD always pull the image from the registry ignoring other caches.

Reviewer 2: Is there a reason for ignoring all caches? This is not necessary given the design of the OCI images. Even if the image has changed it may only have changed the last layer so the previous layers can be re-used from a local cache. According to the image spec that should be the default IIRC. Would be nice to clarify this. Not all implementations have to copy what is used in Kubernetes 1 to 1. Especially if a behavior can cause massive DDOS storms if used incorrectly.

Author: Personally i don't think this means ignore all caches

Reviewer: Coming from the `docker pull` perspective, I read "pull the image from the registry" to mean fetch the manifest for the tag from the registry and if it doesn't match the node cache then pull the layers that diff.

We may need to adjust the wording to make the intention more clear.

* `never`: Compose implementations SHOULD NOT pull the image from a registry and SHOULD rely on the platform cached image. If there is no cached image, a failure MUST be reported.
* `if_not_present`: Compose implementations SHOULD pull the image only if it's not available in the platform cache.This SHOULD be the default option for Compose implementations without build support.

> If `pull_policy` and `build` both presents, Compose implementations SHOULD build the image by default.Compose implementations MAY override this behavior in the tool chain.

Reviewer: Missing a space after a full stop.

```suggestion
If `pull_policy` and `build` both presents, Compose implementations SHOULD build the image by default. Compose implementations MAY override this behavior in the tool chain.
```
